kv: copy key memory before assigning to unreplicated lock

nvb opened the pull request:
Informs #64228.

This commit avoids aliasing between the roachpb.Key assigned to each
unreplicated lock and the bytes returned in the ScanResponse that acquired the
unreplicated lock. It does so by copying keys pulled from ScanResponse memory
before assigning them to unreplicate lock acquisitions.

This is done for two reasons:

1. the keys in an MVCCScanResult, regardless of the scan format used, point
   to a small number of large, contiguous byte slices. These "MVCCScan
   batches" contain keys and their associated values in the same backing
   array. To avoid holding these entire backing arrays in memory and
   preventing them from being garbage collected indefinitely, we copy the key
   slices before coupling their lifetimes to those of unreplicated locks.
2. the KV API has a contract that byte slices returned from KV will not be
   mutated by higher levels. However, we have seen cases (e.g.#64228) where
   this contract is broken due to bugs. To defensively guard against this
   class of memory aliasing bug and prevent keys associated with unreplicated
   locks from being corrupted, we copy them.

Release note (bug fix): large SELECT FOR UPDATE scans will no longer prevent the
memory associated with their entire result set from being reclaimed by the Go
garbage collector for the lifetime of the locks that they acquire.

nvb committed:
kv: copy key memory before assigning to unreplicated lock

Informs #64228.

This commit avoids aliasing between the roachpb.Key assigned to each
unreplicated lock and the bytes returned in the ScanResponse that acquired the
unreplicated lock. It does so by copying keys pulled from ScanResponse memory
before assigning them to unreplicate lock acquisitions.

This is done for two reasons:

1. the keys in an MVCCScanResult, regardless of the scan format used, point
   to a small number of large, contiguous byte slices. These "MVCCScan
   batches" contain keys and their associated values in the same backing
   array. To avoid holding these entire backing arrays in memory and
   preventing them from being garbage collected indefinitely, we copy the key
   slices before coupling their lifetimes to those of unreplicated locks.
2. the KV API has a contract that byte slices returned from KV will not be
   mutated by higher levels. However, we have seen cases (e.g.#64228) where
   this contract is broken due to bugs. To defensively guard against this
   class of memory aliasing bug and prevent keys associated with unreplicated
   locks from being corrupted, we copy them.

Release note (bug fix): large SELECT FOR UPDATE scans will no longer prevent the
memory associated with their entire result set from being reclaimed by the Go
garbage collector for the lifetime of the locks that they acquire.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64428)
<!-- Reviewable:end -->

tbg reviewed (approved):
Do you expect this to produce a noticeable perf hit?

nvb committed:
sql: use GetForUpdate in kvNative.Update

This commit switches `KV/Update/Native` benchmarks to use the GetForUpdate
variant of Get, which acquires an unreplicated lock on the key being queried.
This more closely mirrors how SQL evaluates UPDATE statements.

As expected, unreplicated locks have a cost:
```
➜ benchdiff --count=10 --run='KV/Update/Native' ./pkg/sql/tests
checking out '3fcf931'
building benchmark binaries for '3fcf931' 1/1 -
checking out 'f556c29'
building benchmark binaries for 'f556c29' 1/1 |

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/sql/tests /

name                            old time/op    new time/op    delta
KV/Update/Native/rows=100-16      1.63ms ± 5%    1.83ms ± 3%  +12.24%  (p=0.000 n=10+10)
KV/Update/Native/rows=10-16        287µs ± 4%     324µs ± 6%  +12.66%  (p=0.000 n=10+10)
KV/Update/Native/rows=1-16         132µs ± 6%     149µs ± 6%  +13.25%  (p=0.000 n=10+10)
KV/Update/Native/rows=1000-16     13.3ms ± 3%    15.9ms ± 6%  +19.55%  (p=0.000 n=10+10)
KV/Update/Native/rows=10000-16    98.7ms ± 5%   133.8ms ± 4%  +35.58%  (p=0.000 n=10+9)

name                            old alloc/op   new alloc/op   delta
KV/Update/Native/rows=1-16        20.9kB ± 1%    22.9kB ± 0%   +9.44%  (p=0.000 n=10+10)
KV/Update/Native/rows=10-16       52.9kB ± 0%    65.7kB ± 1%  +24.28%  (p=0.000 n=10+10)
KV/Update/Native/rows=1000-16     3.45MB ± 1%    4.54MB ± 1%  +31.81%  (p=0.000 n=10+10)
KV/Update/Native/rows=100-16       367kB ± 0%     484kB ± 1%  +31.88%  (p=0.000 n=9+10)
KV/Update/Native/rows=10000-16    47.8MB ± 1%    64.2MB ± 2%  +34.34%  (p=0.000 n=10+10)

name                            old allocs/op  new allocs/op  delta
KV/Update/Native/rows=1-16           179 ± 0%       205 ± 0%  +14.53%  (p=0.000 n=10+10)
KV/Update/Native/rows=10-16          381 ± 0%       464 ± 0%  +21.78%  (p=0.000 n=10+10)
KV/Update/Native/rows=10000-16      211k ± 0%      262k ± 0%  +24.32%  (p=0.000 n=9+8)
KV/Update/Native/rows=1000-16      20.8k ± 0%     25.9k ± 0%  +24.72%  (p=0.000 n=9+9)
KV/Update/Native/rows=100-16       2.12k ± 0%     2.67k ± 0%  +25.71%  (p=0.000 n=10+10)
```

nvanbenschoten commented:
> Do you expect this to produce a noticeable perf hit?

It's a good question. I didn't expect anything noticeable, but I was hoping to be able to use the `KV/Update` tests to confirm this. I even added a new commit here to acquire an unreplicated lock in the `KV/Update/Native` benchmark suite. But there was no difference:
```
➜ benchdiff --count=10 --run='KV/Update' ./pkg/sql/tests
checking out '9eae2c2'
building benchmark binaries for '9eae2c2' 1/1 |
checking out '5b06e50'
building benchmark binaries for '5b06e50' 1/1 -

  pkg=1/1 iter=10/10 cockroachdb/cockroach/pkg/sql/tests /

name                            old time/op    new time/op    delta
KV/Update/Native/rows=10000-16     137ms ± 1%     135ms ± 2%  -1.45%  (p=0.008 n=8+9)
KV/Update/Native/rows=1-16         148µs ± 7%     149µs ± 6%    ~     (p=0.529 n=10+10)
KV/Update/Native/rows=10-16        331µs ± 6%     326µs ± 6%    ~     (p=0.393 n=10+10)
KV/Update/Native/rows=100-16      1.88ms ± 5%    1.87ms ± 5%    ~     (p=0.684 n=10+10)
KV/Update/Native/rows=1000-16     16.4ms ± 7%    16.2ms ± 2%    ~     (p=0.356 n=10+9)
KV/Update/SQL/rows=1-16            393µs ± 3%     392µs ± 4%    ~     (p=0.604 n=9+10)
KV/Update/SQL/rows=10-16           623µs ± 4%     623µs ± 5%    ~     (p=0.912 n=10+10)
KV/Update/SQL/rows=100-16         2.14ms ± 7%    2.13ms ± 6%    ~     (p=0.684 n=10+10)
KV/Update/SQL/rows=1000-16        16.3ms ± 6%    16.4ms ± 5%    ~     (p=0.579 n=10+10)
KV/Update/SQL/rows=10000-16        180ms ± 6%     178ms ± 2%    ~     (p=0.481 n=10+10)

name                            old alloc/op   new alloc/op   delta
KV/Update/SQL/rows=100-16          487kB ± 1%     485kB ± 0%  -0.42%  (p=0.019 n=10+10)
KV/Update/Native/rows=10-16       66.0kB ± 1%    65.6kB ± 1%    ~     (p=0.190 n=10+10)
KV/Update/Native/rows=100-16       483kB ± 0%     483kB ± 0%    ~     (p=0.842 n=9+10)
KV/Update/Native/rows=1000-16     4.52MB ± 1%    4.54MB ± 1%    ~     (p=0.123 n=10+10)
KV/Update/Native/rows=10000-16    64.2MB ± 2%    64.2MB ± 1%    ~     (p=0.579 n=10+10)
KV/Update/SQL/rows=1-16           52.1kB ± 0%    52.0kB ± 1%    ~     (p=0.356 n=10+9)
KV/Update/SQL/rows=10-16           108kB ± 0%     108kB ± 0%    ~     (p=0.633 n=10+8)
KV/Update/SQL/rows=1000-16        4.18MB ± 1%    4.18MB ± 1%    ~     (p=0.730 n=9+9)
KV/Update/SQL/rows=10000-16        114MB ± 1%     114MB ± 0%    ~     (p=0.315 n=10+7)
KV/Update/Native/rows=1-16        22.9kB ± 1%    23.1kB ± 1%  +0.71%  (p=0.023 n=9+10)

name                            old allocs/op  new allocs/op  delta
KV/Update/Native/rows=1-16           205 ± 0%       205 ± 0%    ~     (p=0.941 n=9+8)
KV/Update/Native/rows=10-16          463 ± 0%       463 ± 0%    ~     (p=0.471 n=9+9)
KV/Update/Native/rows=100-16       2.67k ± 0%     2.67k ± 0%    ~     (p=1.000 n=10+10)
KV/Update/Native/rows=1000-16      25.9k ± 0%     25.9k ± 0%    ~     (p=0.366 n=10+9)
KV/Update/Native/rows=10000-16      262k ± 0%      262k ± 0%    ~     (p=0.200 n=8+9)
KV/Update/SQL/rows=1-16              516 ± 0%       516 ± 0%    ~     (p=0.181 n=6+10)
KV/Update/SQL/rows=10-16             789 ± 0%       789 ± 0%    ~     (p=0.506 n=9+9)
KV/Update/SQL/rows=100-16          5.86k ± 0%     5.86k ± 0%    ~     (p=0.174 n=9+10)
KV/Update/SQL/rows=1000-16         49.6k ± 0%     49.6k ± 0%    ~     (p=0.810 n=8+10)
KV/Update/SQL/rows=10000-16         594k ± 0%      595k ± 0%    ~     (p=0.133 n=10+9)
```
The reason? As of https://github.com/cockroachdb/cockroach/pull/61583, we now use `GetRequest` instead of `ScanRequest` for most UPDATE statements. But we don't need to copy the key for `GetRequest`, because it comes from the request and not the response. So we'll only hit this in cases where we can't hit that point-read fast-path.

I hacked around that change for the `KV/Update/SQL` suite to continue issuing ScanRequests (effectively reverting #61583) for the sake of experimentation. Here's that result:
```
➜ benchdiff --count=20 --run='KV/Update/SQL' ./pkg/sql/tests
test binaries already exist for '1a363db'; skipping build
test binaries already exist for 'dadb59c'; skipping build

  pkg=1/1 iter=20/20 cockroachdb/cockroach/pkg/sql/tests /

name                         old time/op    new time/op    delta
KV/Update/SQL/rows=1-16         384µs ± 6%     387µs ± 4%    ~     (p=0.247 n=19+20)
KV/Update/SQL/rows=10-16        597µs ± 5%     602µs ± 4%    ~     (p=0.166 n=19+20)
KV/Update/SQL/rows=100-16      2.05ms ± 4%    2.07ms ± 3%    ~     (p=0.314 n=20+20)
KV/Update/SQL/rows=1000-16     16.0ms ± 3%    16.0ms ± 2%    ~     (p=0.762 n=20+18)
KV/Update/SQL/rows=10000-16     175ms ± 4%     175ms ± 5%    ~     (p=0.820 n=20+20)

name                         old alloc/op   new alloc/op   delta
KV/Update/SQL/rows=1-16        52.4kB ± 1%    52.5kB ± 1%    ~     (p=0.890 n=17+18)
KV/Update/SQL/rows=10-16        108kB ± 1%     108kB ± 1%    ~     (p=0.607 n=19+20)
KV/Update/SQL/rows=100-16       485kB ± 1%     485kB ± 1%    ~     (p=0.749 n=19+20)
KV/Update/SQL/rows=1000-16     4.18MB ± 1%    4.18MB ± 1%    ~     (p=0.947 n=20+20)
KV/Update/SQL/rows=10000-16     114MB ± 1%     114MB ± 1%    ~     (p=0.659 n=20+20)

name                         old allocs/op  new allocs/op  delta
KV/Update/SQL/rows=10-16          789 ± 0%       789 ± 0%    ~     (p=0.369 n=20+13)
KV/Update/SQL/rows=100-16       5.86k ± 0%     5.86k ± 0%    ~     (p=0.097 n=19+18)
KV/Update/SQL/rows=1000-16      49.6k ± 0%     49.6k ± 0%    ~     (p=0.769 n=18+19)
KV/Update/SQL/rows=10000-16      595k ± 0%      595k ± 0%    ~     (p=0.476 n=18+18)
KV/Update/SQL/rows=1-16           524 ± 0%       525 ± 0%  +0.17%  (p=0.003 n=19+19)
```
So we see the extra allocation for the case where we acquire an unreplicated lock, but it doesn't have a statistically significant impact on latency.

nvanbenschoten commented:
bors r+
